### ISSUE TITLE
"native:snapgeometries" algorithm doesn't snap vertices in some cases

### ISSUE THREAD

**Issue by blacktesta** [qgis/QGIS#51461]
### What is the bug or the crash?

The title speaks for itself. When I use snapgeometries algorithm it snaps most of vertices but always not every one. I attached sample gpkg for reproducing the problem.
[snap_geometry_test.zip](https://github.com/qgis/QGIS/files/10408289/snap_geometry_test.zip)

### Steps to reproduce the issue

1. Load 2 layers from snap_geometry_test.gpkg
2. Use "native:snapgeometries" algorith for layer_1 with layer_2 as reference layer, tolerance 0.01 ( or even 10) and first behavior (or any other)
3. See result. No new vetices added to layer_1.

### Versions

QGIS version
3.28.2-Firenze
QGIS code revision
b47e00ba601
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.6.1
PROJ version
9.1.1
EPSG Registry database version
v10.076 (2022-08-31)
GEOS version
3.11.1-CAPI-1.17.1
SQLite version
3.39.4
PDAL version
2.4.3
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

Active Python plugins
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

P.S. sorry if I misunderstand something

**Comment by blacktesta** [comment]
After doing some more tests i found that it looks like snapping doesn't work on last segment of ring.

**Comment by gacarrillor** [comment]
I've confirmed this in:
 + QGIS 3.22.7
 + QGIS 3.20.0
 + QGIS 3.16.6
 + QGIS 3.10.10

@nirvn, do you have an idea of why the algorithm has such a strange behavior?

**Comment by lbartoletti** [comment]
> After doing some more tests i found that it looks like snapping doesn't work on last segment of ring.

Not exactly. I think there's a little confusion about what you're asking for and the tool's ambition. Especially as the documentation is very light on this topic.

I guess that you expected two new nodes here (inside the ellipse) :

![image](https://github.com/qgis/QGIS/assets/7521540/3620304a-5f5c-4b36-98a3-d7384ead92ef)

Your geometries are already snapped and aligned, so the algorithm will don't add any new vertices.

If you move the geometry of layer 1, with a new vertex as :

![image](https://github.com/qgis/QGIS/assets/7521540/c2011727-edaa-4c47-8a6b-215b3270e9ab)

the result is :

![image](https://github.com/qgis/QGIS/assets/7521540/f3c216d9-331a-4510-8719-5c82760ca6d4)


The geometry is well snapped and points are aligned.

Maybe you expect a new option to add points from geometries of the second layer?

**Comment by blacktesta** [comment]
@lbartoletti But when i rearrange vertices of polygon on layer_1 algorithm do exactly what i need. See attached GPKG. 
Yes documentation doesn't say much about what exactly this algorithm do.
And plus In my opinion it should have two checkboxes:
1) allow/ do not allow adding vertices
2) allow/ do not allow deleting verticies (because algorithm does it too!)
![image](https://github.com/qgis/QGIS/assets/74047515/a4e5cb29-39a3-4241-998e-bb317957da97)

**Comment by lbartoletti** [comment]
I think we agree together, if we rearrange vertices of polygon on layer_1, the result is more satisfying as you mentioned, and I mentioned - differently - before. However, is not the purpose of these algorithms to add new vertices. BUT, I agree with you, an option to make this job, will be useful. It could be something to add to https://github.com/qgis/QGIS-Enhancement-Proposals/issues/236

Shall I convert this "bug" to a "feature request" ?

**Comment by blacktesta** [comment]
When i choose behavior for this algorithm it says: "Prefer aligning nodes,**insert extra vertices where required**". I understand this like it will move not aligned nodes and **add vertices** where there are none. After this i really dont understand why  adding new vertices is not the purpose of this algorithm. That's why it still seems like a bug to me at this point in time. Correct me if I'm wrong. Despite the fact i got a workaround for this (merge two layers into one, snap to anchor node, split layers back, got new vertices), it would be much easier without this bug.
On the other side i have features to request (like checkboxes I mentioned earlier). I think I should open new "feature request" for these.

**Comment by gacarrillor** [comment]
> @lbartoletti But when i rearrange vertices of polygon on layer_1 algorithm does exactly what i need. See attached GPKG. Yes documentation doesn't say much about what exactly this algorithm do. And plus in my opinion it should have two checkboxes:
> 
>     1. allow/ do not allow adding vertices
> 
>     2. allow/ do not allow deleting verticies (because algorithm does it too!)
> 
> 
> Images with right behavior: ![image](https://user-images.githubusercontent.com/74047515/246340602-a4e5cb29-39a3-4241-998e-bb317957da97.png) [snap_geometry_test_2.zip](https://github.com/qgis/QGIS/files/11767269/snap_geometry_test_2.zip)

Just to add a 3rd opinion to this thread, I've used the **snap geometries to layer** algorithm a lot (the company I worked for in the past even funded performance improvements on it, see [[1](https://github.com/qgis/QGIS/pull/42781)] and [[2](https://github.com/qgis/QGIS/pull/44766)]), and the behavior described by @blacktesta is exactly what I'd expect as a user in this particular case.

**Comment by blacktesta** [comment]
Another bug related to this is if you split lines into segments (2 nodes lines) and try to snap them there will be no points added at all because every line is last segment for itself.

**Comment by blacktesta** [comment]
Hi, @nirvn and thank you again for your inpressive work! I will test it as soon as i can and i will inform you about the results.

**Comment by blacktesta** [comment]
@nirvn I tried build from #56426 with no result. Last segment of a polygon still not snapping.

**Comment by nirvn** [comment]
@blacktesta , OK, thanks for testing. I'll have a look into the specificity of this issue then.

**Comment by blacktesta** [comment]
It's still not working in 2026 on provided layers
